extern_types associated functions aren't visible from other crates

crate `foo`:
```Rust
#![feature(extern_types)]

extern "C" {
    pub type Foo;
}

impl Foo {
    pub fn new() -> FooBox {
        let raw = unimplemented!("Get through ffi");
        FooBox(raw)
    }
}

pub struct FooBox(*mut Foo);

impl Drop for FooBox {
    fn drop(&mut self) {
        unimplemented!("call ffi drop");
    }
}

```

crate `foouser`:
```Rust
extern crate foo;

fn main() {
    let foo = foo::Foo::new();
    println!("{:p}", &foo);
}

```

Error:
```
error[E0599]: no function or associated item named `new` found for type `foo::Foo` in the current scope
 --> foouser.rs:4:15
  |
4 |     let foo = foo::Foo::new();
  |               ^^^^^^^^^^^^^ function or associated item not found in `foo::Foo`
```

This error does not occur if `main` is located within the same crate.

--- a/compiler/rustc_metadata/src/rmeta/encoder.rs
+++ b/compiler/rustc_metadata/src/rmeta/encoder.rs
@@ -1753,6 +1753,7 @@ impl EncodeContext<'a, 'tcx> {
         self.encode_const_stability(def_id);
         self.encode_deprecation(def_id);
         self.encode_item_type(def_id);
+        self.encode_inherent_implementations(def_id);
         if let hir::ForeignItemKind::Fn(..) = nitem.kind {
             record!(self.tables.fn_sig[def_id] <- tcx.fn_sig(def_id));
             self.encode_variances_of(def_id);

--- a/src/test/ui/extern/auxiliary/extern-types-inherent-impl.rs
+++ b/src/test/ui/extern/auxiliary/extern-types-inherent-impl.rs
@@ -1,0 +1,9 @@
+#![feature(extern_types)]
+
+extern "C" {
+    pub type CrossCrate;
+}
+
+impl CrossCrate {
+    pub fn foo(&self) {}
+}

--- a/src/test/ui/extern/extern-types-inherent-impl.rs
+++ b/src/test/ui/extern/extern-types-inherent-impl.rs
@@ -1,19 +1,26 @@
-// run-pass
-#![allow(dead_code)]
 // Test that inherent impls can be defined for extern types.
+
+// check-pass
+// aux-build:extern-types-inherent-impl.rs
 
 #![feature(extern_types)]
 
-extern {
-    type A;
+extern crate extern_types_inherent_impl;
+use extern_types_inherent_impl::CrossCrate;
+
+extern "C" {
+    type Local;
 }
 
-impl A {
-    fn foo(&self) { }
+impl Local {
+    fn foo(&self) {}
 }
 
-fn use_foo(x: &A) {
+fn use_foo(x: &Local, y: &CrossCrate) {
+    Local::foo(x);
     x.foo();
+    CrossCrate::foo(y);
+    y.foo();
 }
 
-fn main() { }
+fn main() {}

